File: extension/content/content.ml
open Brr
open Fut.Syntax
module Runtime = Chrome.Runtime

let scrape_page () =
  let open Brr in
  let title = Document.title G.document |> Jv.of_jstr in
  let body = Document.body G.document |> El.to_jv in
  let inner_text = Jv.get body "innerText" in
  (title, inner_text)

let send runtime request send_response =
  let message_type = ("type", Jv.of_string "response") in
  let message_action = ("action", Jv.of_string "save") in
  let+ send_fut = Runtime.send_message request runtime in
  match send_fut with
  | Error err ->
      let result = ("result", Jv.of_jstr (Jv.Error.message err)) in
      Jv.apply send_response [| Jv.obj [| message_type; message_action; result |] |]
  | Ok res ->
      let result = ("result", res) in
      Jv.apply send_response [| Jv.obj [| message_type; message_action; result |] |]

let handle_requests runtime request _sender send_response =
  let title, inner_text = scrape_page () in
  let payload = Jv.get request "payload" in
  Jv.set payload "title" title;
  Jv.set payload "body" inner_text;
  let _ = send runtime request send_response in
  Jv.of_bool true

let listener runtime request _sender send_response =
  match Jv.get request "type" |> Jv.to_string with
  | "ping" ->
      let response = Jv.obj [| ("type", Jv.of_string "pong") |] in
      ignore (Jv.apply send_response [| response |]);
      Jv.of_bool true
  | _ -> handle_requests runtime request _sender send_response

let () =
  let runtime = Chrome.runtime in
  Runtime.(runtime |> on_message |> On_message.add_listener (listener runtime));
  Console.(log [ str "Noematic scrape handler installed" ])

File: extension/content/content.js
/**
 * @typedef {import('../common/types.js').Responder} Responder
 */

/**
 * @param {any} request
 * @param {chrome.runtime.MessageSender} _sender
 * @param {Responder} sendResponse
 * @returns {boolean}
 */
const handleSaveRequests = (request, _sender, sendResponse) => {
  request.payload.innerText = document.body.innerText;
  chrome.runtime.sendMessage(request).then((response) => {
    response.action = 'saveResponse';
    sendResponse(response);
  });
  return true;
};

/**
 * @param {any} _request
 * @param {chrome.runtime.MessageSender} _sender
 * @param {Responder} sendResponse
 * @returns {boolean}
 */
const handlePingRequests = (_request, _sender, sendResponse) => {
  sendResponse({ action: 'pong' });
  return true;
};

/**
 * @param {any} request
 * @param {chrome.runtime.MessageSender} sender
 * @param {Responder} sendResponse
 * @returns {boolean | undefined}
 */
const onMessageListener = (request, sender, sendResponse) => {
  switch (request.action) {
    case 'saveRequest':
      return handleSaveRequests(request, sender, sendResponse);
    case 'ping':
      return handlePingRequests(request, sender, sendResponse);
    default:
      return undefined;
  }
};

chrome.runtime.onMessage.addListener(onMessageListener);

console.log('Noematic scrape handler installed');
